platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/GraphVariantSelector.java

```diff
@@ -158,7 +158,7 @@ public VariantGraphResolveState selectLegacyVariant(ImmutableAttributes consumer
             // We wanted to do variant matching, but there were no variants in the target component.
             // So, we fell back to looking for the legacy (`default`) configuration, but it didn't exist.
             // So, there are no variants to select from, and selection fails here.
-            throw failureHandler.noVariantsExistFailure(consumerSchema, consumerAttributes, targetComponentState.getId());
+            throw failureHandler.noCompatibleVariantsFailure(consumerSchema, consumerSchema.matcher(), targetComponentState, consumerAttributes, ImmutableCapabilities.EMPTY, targetComponentState.getCandidatesForGraphVariantSelection());
```

Comment:
I think this is the wrong change. 

`noCompatibleVariantsFailure` is for when we tried to select via attribute matching, but none were compatible with our request attributes. 

We get here when there are _no_ target variants. Thus, we look for the default variant (which may not be in that list of target variants), and that does not exist either. 

When we get here, no variants exist.

Reply:
Well, if no variants exist, then obviously no compatible variants exist. QED. 😄 

But, yeah, I struggled a little with what to throw here, too.

The `NoCompatibleVariantsFailureDescriber` actually handles 3 cases: no variants, no variants with attributes, and no matching variants.  But there is only a single, more general `noCompatibleVariantsFailure` method on the failure handler.  

I don't want to pull apart the describer in this PR, but I can add a `noVariantsFailure` method and a note to split apart this case.